src/Workspaces/Core/Portable/Utilities/ParameterName.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System;
+using System.Collections.Generic;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Utilities
+{
+    internal struct ParameterName : IEquatable<ParameterName>
+    {
+        //public static readonly IEqualityComparer<ParameterName> OrdinalComparer = new ParameterNameEqualityComparer(StringComparer.Ordinal);
+        //public static readonly IEqualityComparer<ParameterName> OrdinalIgnoreCaseComparer = new ParameterNameEqualityComparer(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// The name the underlying naming system came up with based on the argument itself.
+        /// This might be a name like "_value".  We pass this along because it can help
+        /// later parts of the GenerateConstructor process when doing things like field hookup.
+        /// </summary>
+        public readonly string NameBasedOnArgument;
+
+        /// <summary>
+        /// The name we think should actually be used for this parameter.  This will include
+        /// stripping the name of things like underscores.
+        /// </summary>
+        public readonly string BestNameForParameter;
+
+        public ParameterName(string nameBasedOnArgument, bool isFixed)
+        {
+            NameBasedOnArgument = nameBasedOnArgument;
+
+            if (isFixed)
+            {
+                // If the parameter name is fixed, we have to accept it as is.
+                BestNameForParameter = NameBasedOnArgument;
+            }
+            else
+            {
+                // Otherwise, massage it a bit to be a more suitable match for
+                // how people actually writing parameters.
+                var trimmed = nameBasedOnArgument.TrimStart('_');
+                BestNameForParameter = trimmed.Length > 0 ? trimmed.ToCamelCase() : nameBasedOnArgument;
+            }
+        }
+
+        public override bool Equals(object obj)
+        {
+            return Equals((ParameterName)obj);
```

Comment:
The normal pattern is `obj as ParameterName` so that this doesn't throw if you compare against something else.